ocaml/xapi/helpers.ml

```diff
@@ -1230,3 +1230,40 @@ let host_supports_hvm ~__context host =
 let env_with_path env_vars =
   Array.of_list @@
   List.map (fun (k,v) -> Printf.sprintf "%s=%s" k v) (("PATH", String.concat ":" Forkhelpers.default_path)::env_vars)
+
+let ensure_same_protocol (ips: Ipaddr.t list) =
+  let coerce_v4 xs = xs |> List.map (function
+    | Ipaddr.V4 x -> x
+    | V6        x -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "expected IPv4, but got an IPv6: %s" (Ipaddr.V6.to_string x)])))
+  in
+  let coerce_v6 xs = xs |> List.map (function
+    | Ipaddr.V6 x -> x
+    | V4        x -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "expected IPv6, but got an IPv4: %s" (Ipaddr.V4.to_string x)])))
+  in
+  match ips with
+  | []                  -> failwith "expected at least one ip"
+  | Ipaddr.V4 ip :: ips -> `Ip_v4s (ip::coerce_v4 ips)
+  | V6        ip :: ips -> `Ip_v6s (ip::coerce_v6 ips)
+
+let intersect_ips_exn ips =
+  let and_ipv6 x y =
+    let (x_ho, x_lo) = Ipaddr.V6.to_int64 x in
+    let (y_ho, y_lo) = Ipaddr.V6.to_int64 y in
+    Ipaddr.V6.of_int64 (Int64.logand x_ho y_ho, Int64.logand x_lo y_lo)
+  in
+  let and_ipv4 x y = Int32.logand (Ipaddr.V4.to_int32 x) (Ipaddr.V4.to_int32 y) |> Ipaddr.V4.of_int32 in
+  match ensure_same_protocol ips with
+  | `Ip_v4s (ip::ips) -> Ipaddr.V4 (List.fold_left and_ipv4 ip ips)
+  | `Ip_v6s (ip::ips) -> Ipaddr.V6 (List.fold_left and_ipv6 ip ips)
+  | `Ip_v4s [] | `Ip_v6s [] -> failwith "impossible"
```

Comment:
Minor comment: this case is not impossible in the sense that the program can never reach this state. Instead, you assume the list of IPs is never empty. I think a better error would be something like `list of IPs must not be empty` or similar.